> In the past I've encountered similar issues where upstream was using some other kind of build-system, thus not creating that folder. At the time they also provided a `setuptools` method for building, so I used that instead.
> 
> In your case I suppose `bazel` must have a way to create such a folder when install the artifact (a `.whl`?) with `pip`.

Yes, I can use `bazel` to create a wheel and then install that wheel with `buildPythonPackage` as their official install-from-source instruction suggest:

```bash
git clone https://github.com/tensorflow/probability.git
cd probability
bazel build --copt=-O3 --copt=-march=native :pip_pkg
PKGDIR=$(mktemp -d)
./bazel-bin/pip_pkg $PKGDIR
pip install --user --upgrade $PKGDIR/*.whl
```

But doing that, `pip` will for some reason try to replace `tensorflow` (which is provided as a dependency):

```
Requirement already satisfied: numpy>=1.13.3 in /nix/store/3ml16mnq11in68sdg8w4jipm7bqcvmh8-python3.7-numpy-1.16.4/lib/python3.7/site-packages (from tensorflow==probability)>
Installing collected packages: tensorflow
  Found existing installation: tensorflow 1.13.1
    Uninstalling tensorflow-1.13.1:
ERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/nix/store/5vs1nfppwaigihl3ada4d68fj7ffxzdi-python3.7-tensorflow-1.13.1/bin/free>
Consider using the `--user` option or check the permissions.
```

If I remove `tensorflow` from `propagatedBulidInputs` it instead complains about `.dist-info`:

```
Installing collected packages: tensorflow
ERROR: Exception:
Traceback (most recent call last):
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/cli/base_command.py", line 178, in main
    status = self.run(options, args)
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/commands/install.py", line 414, in run
    use_user_site=options.use_user_site,
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/req/__init__.py", line 58, in install_given_r>
    **kwargs
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/req/req_install.py", line 920, in install
    use_user_site=use_user_site, pycompile=pycompile,
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/req/req_install.py", line 448, in move_wheel_>
    warn_script_location=warn_script_location,
  File "/nix/store/pwqjs8nz67h4rid05vwma9z8fbsfmm16-python3.7-bootstrapped-pip-19.1.1/lib/python3.7/site-packages/pip/_internal/wheel.py", line 428, in move_wheel_files
    assert info_dir, "%s .dist-info directory not found" % req
AssertionError: tensorflow==probability .dist-info directory not found
```

So the wheel is created, but the .dist-info dir apparently not. Also the part where it tries to replace `tensorflow` is curious. Any idea why?

Here's a link to their setup.py: https://github.com/tensorflow/probability/blob/master/setup.py

--- a/pkgs/build-support/build-bazel-package/default.nix
+++ b/pkgs/build-support/build-bazel-package/default.nix
@@ -88,8 +88,8 @@ in stdenv.mkDerivation (fBuildAttrs // {
   '';
 
   preConfigure = ''
-    mkdir -p $bazelOut/external
-    cp -r $deps/* $bazelOut/external
+    mkdir -p "$bazelOut"
+    cp -r $deps $bazelOut/external
     chmod -R +w $bazelOut
     find $bazelOut -type l | while read symlink; do
       ln -sf $(readlink "$symlink" | sed "s,NIX_BUILD_TOP,$NIX_BUILD_TOP,") "$symlink"

--- a/pkgs/build-support/build-bazel-package/default.nix
+++ b/pkgs/build-support/build-bazel-package/default.nix
@@ -51,8 +51,10 @@ in stdenv.mkDerivation (fBuildAttrs // {
       rm -rf $bazelOut/external/{local_*,\@local_*}
 
       # Patching markers to make them deterministic
-      sed -i 's, -\?[0-9][0-9]*$, 1,' $bazelOut/external/\@*.marker
-      sed -i '/^ENV:TMP.*/d' $bazelOut/external/\@*.marker
+      find $bazelOut/external -name '@*\.marker' -exec sed -i \
+        -e 's, -\?[0-9][0-9]*$, 1,' \
+        -e '/^ENV:TMP.*/d' \
+        '{}' \;
 
       # Remove all vcs files
       rm -rf $(find $bazelOut/external -type d -name .git)

--- a/pkgs/development/python-modules/tensorflow-probability/default.nix
+++ b/pkgs/development/python-modules/tensorflow-probability/default.nix
@@ -1,8 +1,13 @@
 { lib
 , fetchFromGitHub
+, buildBazelPackage
 , buildPythonPackage
 , python
+, setuptools
+, wheel
 , tensorflow
+, six
+, numpy
 , decorator
 , cloudpickle
 , hypothesis
@@ -12,19 +17,62 @@
 , pytest
 }:
 
-buildPythonPackage rec {
-  pname = "tensorflow-probability";
+let
   version = "0.7";
+  pname = "tensorflow_probability";
 
-  src = fetchFromGitHub {
-    owner = "tensorflow";
-    repo = "probability";
-    rev = "v${version}";
-    sha256 = "0sy9gmjcvmwciamqvd7kd9qw2wd7ksklk80815fsn7sj0wiqxjyd";
+  # first build all binaries and generate setup.py using bazel
+  bazel-wheel = buildBazelPackage {
+    name = "${pname}-${version}-py2.py3-none-any.whl";
+
+    src = fetchFromGitHub {
+      owner = "tensorflow";
+      repo = "probability";
+      rev = "v${version}";
+      sha256 = "0sy9gmjcvmwciamqvd7kd9qw2wd7ksklk80815fsn7sj0wiqxjyd";
+    };
+
+    nativeBuildInputs = [
+      # needed to create the output wheel in installPhase
+      python
+      setuptools
+      wheel
+    ];
+
+    bazelTarget = ":pip_pkg";
+
+    fetchAttrs = {
+      sha256 = "0sjjj9z1dhilhpc8pq4154czrb79z9cm044jvn75kxcjv6v5l2m5";
+    };
+
+    buildAttrs = {
+      preBuild = ''
+        patchShebangs .
+      '';
+
+      installPhase = ''
+        # work around timestamp issues
+        # https://github.com/NixOS/nixpkgs/issues/270#issuecomment-467583872
+        export SOURCE_DATE_EPOCH=315532800
+
+        # First build, then move. Otherwise pip_pkg would create the dir $out
+        # and then put the wheel in that directory. However we want $out to
+        # point directly to the wheel file.
+        ./bazel-bin/pip_pkg . --release
+        mv *.whl "$out"
+      '';
+    };
   };
+in buildPythonPackage rec {
+  inherit version pname;
+  format = "wheel";
+
+  src = bazel-wheel;
 
   propagatedBuildInputs = [
     tensorflow
+    six
+    numpy
     decorator
     cloudpickle
   ];

